docs/content/sips/006-spin-plugins.md

```diff
@@ -0,0 +1,233 @@
+
+```bash
+Install a Spin plugin using a plugin manifest file. 
+By default, looks for the plugin manifest named <name>.json
+in the Spin plugins repository https://github.com/fermyon/spin-plugins
+
+USAGE:
+    spin plugin install <name>
+
+OPTIONS:
+    -f, --file                       Path to local plugin manifest
+	-u, --url                        Address of remote plugin manifest
+    -v, --version                    Desired version to update plugin to. Defaults to latest.
+    -s, --skip-version-compat-check  Instructs Spin to install the plugin even 
+                                     if its manifest specifies that it is 
+                                     incompatible with the current version of Spin
+```
+
+If the manifest is found, Spin will check that the plugin is compatible with the current OS, platform, and version of Spin. If so, before installing the plugin, Spin will prompt the user as to whether to trust the source. For example, the following prompt would be displayed for a plugin named `deploy` with an Apache 2 license and hosted at [`https://github.com/fermyon/spin-plugin-deploy/releases/download/v0.1.0/spin-plugin-deploy-v0.1.0-macos-aarch64.tar.gz`](https://github.com/fermyon/spin-plugin-deploy/releases/download/v0.1.0/spin-plugin-deploy-v0.1.0-macos-aarch64.tar.gz):
+
+```bash
+Installing plugin deploy with license Apache 2.0 from https://github.com/fermyon/spin-plugin-deploy/releases/download/v0.1.0/spin-plugin-deploy-v0.1.0-macos-aarch64.tar.gz
+For more information, reference the plugin metadata at `https://github.com/fermyon/spin-plugins/plugin-manifests/deploy.json`.
+Are you sure you want to proceed? ('yes'/'no') (default: no)
+```
+
+The plugin will only be installed if a user enters `yes`. Otherwise, the command exits. 
+
+Spin will reference the plugin manifest in order to fetch the plugin binary and install it into the user’s local data directory under a Spin-managed `plugins` subdirectory. The plugin manifest will be stored within a `manifests` subdirectory. 
+
+After installing a plugin, it can be executed directly from the Spin CLI. For example, a plugin named `$name` would be executed by running `spin $name <args>`. Any additional arguments supplied will be passed when executing the associated binary.
+
+ 
+
+**`spin plugin uninstall`**
+
+The `spin plugin uninstall` command uninstalls a plugin named `$name`. 
+
+```bash
+Uninstall a Spin plugin.
+
+USAGE:
+    spin plugin uninstall <name>
+```
+
+**`spin plugin update`**
+
+The `spin plugin update` command updates one or all plugins. If updating a single plugin, the desired version can be specified. By default, plugins are updated to the latest version in the plugins repository. As with `spin plugin install`, the local path or remote addresses to a plugin manifest can be specified.
+
+```bash
+Update one or all installed Spin plugins.
+
+USAGE:
+    spin plugin update [OPTIONS]
```

Comment:
Is it worth introducing a third command called `spin plugin upgrade` to handle this? Kinda like how you have to `apt update` your repositories before you `apt upgrade`. Wondering what might be the implications involved by compressing both steps into one command.

Reply:
Are you proposing breaking this down into a "check if updates exist" and "do update" step? We aren't caching available plugin information so i don't see the need for the former. What benefit to you see just the update side having?

Reply:
Yes.

Some form of caching/update lookup mechanism may be necessary if we are to use GitHub repos for indexing/searching. I know homebrew and other package managers got into trouble from causing tons of strain upstream due to commands like `brew search` cloning github.com/Homebrew/homebrew-core on every call...

If the plan is to implement a `spin plugin search` from the spin-plugins repo, it may be best to store a local cache to prevent unnecessary load upstream.

Might be putting the cart before the horse a bit, but, food for thought :)

Reply:
We could just cache the repo locally on `spin plugin update`, which calls to `spin plugin search` could use. But then users wouldn't be able to update their search results until they `spin plugin update` which may cause their plugins to update. That may or may not be intended behaviour, but I could see both arguments:

1. I just wanted to search for new plugins, not update them!
2. Oh hey! New updates came out and now I'm on the latest. Whee!

Reply:
I think it would make sense to split it into `update` and `upgrade` if we are going to be caching the plugins repo to avoid unexpected upgrades for the user.

Reply:
These are all great points! And now i understand why cloning the entire repo is the optimal path of fetching plugins. Looks like we have 3 options (i may be missing some):
1. Add `spin plugin update` that fetches latest spin repo contents and rename current command to `spin plugin upgrade`
1. Just do latter of above, renaming `spin plugin update` to `spin plugin upgrade`. Always fetch latest before upgrades. Later, can add a separate `spin plugin update` that fetches the latest spin-repo contents. 
1. Add a `--fetch-latest` flag to `spin plugin update` (and later `spin plugin search`) that adds a preliminary step of pulling the latest spin-repo contents. This will avoid the scenario you gave of having to upgrade in order to search latest.

Reply:
I like option 2 the best. Seems like the simplest option (just rename the command) while keeping the door open for future enhancements (`spin plugin update`).

Reply:
Thanks for this recommendation @bacongobbler. Renamed in 76aa5da02f74b